Integrate with encoding TransferSyntax

pevers opened the pull request:
Hi @Enet4 ,

I think issue https://github.com/Enet4/dicom-rs/issues/120 is a great start to start supporting native Rust decoding/encoding of pixel data. I see that JPEG lossless is being added to jpeg-decoder here: https://github.com/image-rs/jpeg-decoder/issues/191 . And RLE and Deflate are implemented here: https://github.com/rusty-shell/rust-compress. If we have this abstraction we could easily implement those and not have to rely on GDCM. Supporting multi-frame dicoms would then also be easier. After that, we can port J2K, JPEG-LS etc.

This PR is a far from ready. I hope that we can discuss this issue and see what works. If it is totally off course I can just discard it. I was a bit confused about the `DataRWAdapter`. I tried using this as a trait object to decode/encode pixel data as well. But then it is hard to use because the reader is a Box with a static lifetime. So I simplified it a bit to use the `PixelRWAdapter` instead.

```rust
(*decoder).decode(fragments[0].as_slice())
```

I also split the pixel data decoder from the `Codec`. But I don't think that is necessary. I did it for now so that it was easier to make some adjustments, but I think it can just become:

```
pub enum Codec<A, B> {
...,
PixelData(B)
}
```

If you have the time. Can you tell me I'm heading in the right direction? If not, or if you want to do this yourself, please don't hesitate to close the PR ;)

Enet4 reviewed (commented):
Hello @pevers! I greatly appreciate your initiative here!

I generally liked the way this is being developed, and we should be able to reuse most of it. It does raise food for thought, and I have already left some comments below for further guidance. It is currently unclear whether we will face problems related with these matters, but this is a great time to reflect on these problems. :)

pevers commented on encoding/src/adapters/mod.rs:
```diff
@@ -0,0 +1,27 @@
+pub mod rle_lossless;
+pub trait PixelRWAdapter {
+  /// Decode byte stream
+  fn decode(&self, reader: &[u8]) -> Vec<u8>;
+
+  /// Write byte stream
+  fn encode(&self, writer: &[u8]) -> Vec<u8>;
```
You are right! This interface is way too simple. I'll implement a basic example using RLE Lossless to see what works.

I think for decoding we should maybe accept a `PixelSequence`. We should also support decoding frame-by-frame. Otherwise it would be a bad experience decoding a 700MB dicom.

So maybe something like this:

```rust
#[derive(Debug, Clone, PartialEq)]
pub struct PixelSequence<'a> {
    pub offset_table: &'a C<u8>,
    pub fragments: &'a C<Vec<u8>>
}
```

I'm trying to losen the constraints as much as possible without having to copy pixel data when it is not needed.

And then decode it:

```rust
fn decode(&self, src: &PixelSequenceType, frame: u16, dst: &mut Vec<u8>) -> DecodeResult<()>;
```

For encoding I still need to come up with a solution.

Enet4 replied:
> So maybe something like this:

Passing that `PixelSequence` value to the decoding process would still require the whole raw pixel data to exist in memory. Although this is currently true for the existing DICOM object implementation (`InMemoryDicomObject`), it will not be the case for the upcoming lazy DICOM object.

Looking back to some loose ideas on this subject, I think the original intent of having this new `PixelDataObject` trait was so that a trait object could be passed to the `decode` method. Let's try this:

```rust
fn decode(&self, src: &dyn PixelDataObject, frame: u16, dst: &mut Vec<u8>) -> DecodeResult<()>;
```

On a minor side note, wouldn't `u32` be a more suitable type for the frame number?

pevers commented on encoding/src/transfer_syntax/mod.rs:
```diff
@@ -398,7 +401,7 @@ impl<A> TransferSyntax<A> {
             byte_order: self.byte_order,
             explicit_vr: self.explicit_vr,
-            codec,
+            codec
```
Thanks! Yeah I didn't setup any proper hooks. I now have some hooks setup that trigger when I commit :D

pevers commented on pixeldata/src/lib.rs:
```diff
@@ -400,6 +403,40 @@ where
+pub trait PixelDataObject {
+    // TODO: And much more ...
+
+    fn raw_pixel_data(&self) -> Result<Vec<u8>>;
```
Yes, that is exactly the problem I ran into. Nice read in the RFC. Thanks for pointing me to this: https://github.com/rust-lang/rfcs/blob/master/text/1598-generic_associated_types.md

Enet4 replied:
Right, maybe I should have provided a link before. The tracker issue for the feature is rust-lang/rust#44265. Years of waiting will hopefully be worth it one day.

pevers replied:
@Enet4 talking of the devil 😀 https://blog.rust-lang.org/2021/08/03/GATs-stabilization-push.html

pevers commented on pixeldata/src/lib.rs:
```diff
@@ -57,10 +57,13 @@
 
 use byteorder::{ByteOrder, NativeEndian};
 use dicom_core::{value::Value, DataDictionary};
+use dicom_encoding::adapters::DecodeError;
```
These unused warnings can be fixed by adding `#[cfg(not(feature = "gdcm"))]` . But it might pollutes the imports too much and since we get rid of GDCM I didn't include it.

pevers commented:
Hi there @Enet4 ,

I tried to tidy up this PR so that it is ready for a review.

This is a summary of what I did:

1. Added native RLE decoding support (no encoding yet). The decoder selectively reads fragments since the TS defines a one-to-one relationship between fragments and frames. This is different depending on the TS. I know for example that for Lossless JPEG, GDCM reads in EOF markers to indicate a frame start.

1. Added multi-frame support. I changed the `to_dynamic_image` method by adding a frame number.

1. Removed applying Rescale Slope and Intercept for now. It doesn't make sense to randomly apply it without applying VOI LUT. We can pick that up in https://github.com/Enet4/dicom-rs/issues/122 .

It is still required to decode the whole image to display only 1 frame. This is also what Pydicom is doing and I think it is OK. The decoder is able to selectively read the pixel data from fragments, so that it doesn't hold another copy of all pixel data in memory. 

Thanks for taking the time to review!

Enet4 commented on encoding/src/adapters/rle_lossless.rs:
```diff
@@ -0,0 +1,198 @@
+use super::{CustomDecodeError, MissingAttribute};
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct RLELosslessAdapter {}
```
Can be a unit struct.
```suggestion
pub struct RLELosslessAdapter;
```

Enet4 reviewed (changes requested):
Thanks, @pevers! I left a few code review comments to further fine tune these changes.

> This is a summary of what I did:
>
>  1. Added native RLE decoding support (no encoding yet). The decoder selectively reads fragments since the TS defines a one-to-one relationship between fragments and frames. This is different depending on the TS. I know for example that for Lossless JPEG, GDCM reads in EOF markers to indicate a frame start.

That is accurate. The part where either the decoder or the encoder may be present is intriguing, and I wonder whether it is worth encoding that in the type system. But I suppose that the decoder is the most critical one, considering that applications which do not want a given transfer syntax will require the other application to decode the pixel data.

>
> 2. Added multi-frame support. I changed the to_dynamic_image method by adding a frame number.

Sounds good!

>
> 3. Removed applying Rescale Slope and Intercept for now. It doesn't make sense to randomly apply it without applying VOI LUT. We can pick that up in [pixeldata] API extension for selective application of Modality and VOI LUT functions #122 .

This sounds like the kind of operation which would be nice to have, even if no window levels are applied afterwards. But I agree to do a better handling of this with #122.

>
> It is still required to decode the whole image to display only 1 frame. This is also what Pydicom is doing and I think it is OK. The decoder is able to selectively read the pixel data from fragments, so that it doesn't hold another copy of all pixel data in memory.

Yes, this should be fine.

pevers commented:
I believe that I fixed all feedback. Thanks again for the help.

I created an example here: https://peterevers.nl/dicom-rs-wasm/ . That showcases the WASM support. As you can see the planar configuration is not yet supported for the RLE encoded Dicom. I'll create an issue for that.
